File: Task1C/task1c/src/main/java/com/cps2004/Clothes.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public abstract class Clothes extends Product{
    private String size;
    private String brand;
    private String material;

    // Default constructor
    public Clothes(){
        super();
        this.size = "";
        this.brand = "";
        this.material = "";
    }

    // Parameterized constructor
    public Clothes(String name, int id, Supplier supplier, int price, int volume, String size, String brand, String material){
        super(name, id, supplier, price, volume);
        this.size = size;
        this.brand = brand;
        this.material = material;
    }

    // Getters and setters
    public String getSize() { return size; }
    public String getBrand() { return brand; }
    public String getMaterial() { return material; }
    public void setSize(String size) { this.size = size; }
    public void setBrand(String brand) { this.brand = brand; }
    public void setMaterial(String material) { this.material = material; }

    @Override
    // Calculate discounted price
    public abstract double calculate_discount(String month, int quantity);

    // Display product information
    public void display_product_info() {
        super.display_product_info();
        System.out.println("Product size: " + size);
        System.out.println("Product brand: " + brand);
        System.out.println("Product material: " + material);
    }

    // Save product information to file
    public void save(ObjectNode newItem) {
        super.save(newItem);
        newItem.put("size", this.size);
        newItem.put("brand", this.brand);
        newItem.put("material", this.material);
    }

    // Load product information from file
    public void load(JsonNode node) {
        super.load(node);
        size = node.get("size").asText();
        brand = node.get("brand").asText();
        material = node.get("material").asText();
    }
}

File: Task1B/clothes.cpp
#include "clothes.h"

// Default constructor
Clothes::Clothes() {
    this->size = "";
    this->brand = "";
    this->material = "";
}

// Parameterised constructor
Clothes::Clothes(string name, int price, Supplier* supplier, int volume, int productID, string size, string brand, string material) : Product(name, price, supplier, volume, productID) {
    this->size = size;
    this->brand = brand;
    this->material = material;
}

// Getters and setters
string Clothes::getSize() {
    return this->size;
}

string Clothes::getBrand() {
    return this->brand;
}

string Clothes::getMaterial() {
    return this->material;
}

void Clothes::setSize(string size) {
    this->size = size;
}

void Clothes::setBrand(string brand) {
    this->brand = brand;
}

void Clothes::setMaterial(string material) {
    this->material = material;
}

// Display product information
void Clothes::display_product_info() {
    Product::display_product_info();
    cout << "Size: " << this->size << endl;
    cout << "Brand: " << this->brand << endl;
    cout << "Material: " << this->material << endl;
}

// Save product information to file
void Clothes::save(ostream& out) const {
    Product::save(out);
    out << size << "," << brand << "," << material << "\n";
}

// Load product information from file
void Clothes::load(istream& in) {
    Product::load(in);

    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    size = token;

    getline(ss, token, ',');
    brand = token;
    
    getline(ss, token, ',');
    material = token;
}